src/lib.rs

```diff
@@ -356,33 +370,97 @@ impl KaboodleInner {
                     let mut known_peers = self.known_peers.lock().await;
-                    let now = Instant::now();
+                    let timestamp = Instant::now().checked_sub(MAX_PEER_SHARE_AGE).unwrap();
                     for peer in peers {
-                        known_peers.entry(peer).or_insert(PeerState::Known(now));
+                        known_peers
+                            .entry(peer)
+                            .or_insert(PeerState::Known(timestamp));
                     }
                     drop(known_peers);
                 }
-                SwimMessage::Ping => {
-                    self.send_msg(&sender, &SwimMessage::Ack(self.self_addr))
+                SwimMessage::KnownPeersRequest(their_fingerprint, their_num_peers) => {
+                    let mut known_peers = self.known_peers.lock().await;
+                    known_peers.insert(sender, PeerState::Known(Instant::now()));
+                    let our_fingerprint =
+                        generate_fingerprint(&known_peers)[0..SHORT_FINGERPRINT_LENGTH].to_owned();
+                    let our_num_peers = known_peers.len() as u32;
+
+                    // Send back a list of every other peer (besides ourselves and the requestor)
+                    // who is in the Known state. We are specifically excluding peers in the
+                    // WaitingFor... states because otherwise, we are more likely to accidentally
+                    // propagate echoes of down-but-not-yet-noticed-down nodes.
+                    let other_peers = known_peers
+                        .iter()
+                        .filter(|(other_peer, other_state)| {
+                            **other_peer != self.self_addr
+                                && **other_peer != sender
+                                && match other_state {
+                                    PeerState::Known(ts) => {
+                                        Instant::now().duration_since(*ts) < MAX_PEER_SHARE_AGE
+                                    }
+                                    _ => false,
+                                }
+                        })
+                        .map(|(other_peer, _)| other_peer.to_owned())
+                        .collect();
+                    drop(known_peers);
+                    self.send_msg(&sender, &SwimMessage::KnownPeers(other_peers))
                         .await;
+
+                    self.maybe_sync_known_peers(
+                        sender,
+                        our_fingerprint,
+                        our_num_peers,
+                        their_fingerprint,
+                        their_num_peers,
+                    )
+                    .await;
+                }
+                SwimMessage::Ping => {
+                    let known_peers = self.known_peers.lock().await;
+                    let our_fingerprint = generate_fingerprint(&known_peers);
+                    let our_num_peers = known_peers.len() as u32;
```

Comment:
I refactored `maybe_sync_known_peers` to just take the peer address, their fingerprint, and peer count—it recalculates our fingerprint and peer count. It's a bit more work (we have to acquire the lock) but it simplifies the call sites.

As a bonus, it made me notice a bug, which is that our response to  indirect pings on behalf of other peers was sending back _our_ fingerprint and peer count, rather than the ping target's fingerprint and peer count. Whoops.